migrations/2024-12-18-200602_optimize_get_random_community/up.sql

```diff
@@ -0,0 +1,20 @@
+-- * `smallint` range from https://www.postgresql.org/docs/17/datatype-numeric.html
+-- * built-in `random` function has `VOLATILE` and `PARALLEL RESTRICTED` according to:
+--   * https://www.postgresql.org/docs/current/parallel-safety.html#PARALLEL-LABELING
+--   * https://www.postgresql.org/docs/17/xfunc-volatility.html
+CREATE FUNCTION random_smallint ()
+    RETURNS smallint
+    LANGUAGE sql
+    VOLATILE PARALLEL RESTRICTED RETURN
+    -- https://stackoverflow.com/questions/1400505/generate-a-random-number-in-the-range-1-10/1400752#1400752
+    -- (`+ 1` makes the upper bound inclusive)
+    trunc ((random() * (32768 + 32767 + 1)) - 32768
+);
+
+ALTER TABLE community
+    ADD COLUMN random_number smallint NOT NULL DEFAULT random_smallint ();
+
+CREATE INDEX idx_community_random_number ON community (random_number) INCLUDE (local, nsfw)
+WHERE
+    NOT (deleted OR removed OR visibility = 'Private');
+
```

Comment:
why smallint? due to birthday paradox you have a high collision probability already with 2^(n/2) communities = 2^8 = 256 communities

Reply:
It uses less space, and collisions are fine because the random number is updated after the community is picked

Reply:
why would you update the number after picking?

Reply:
Its [described here](https://github.com/LemmyNet/lemmy/pull/5267/files#diff-f8ab24cb9e7a0ea0a409442a1da80bb9aea619911c207e07b42326004eaa754dR217). I'm sure there are others ways of doing it, allowing collisions and updating the random number after picking, being one way to do it.

Reply:
right, i missed the last paragraph there. Still seems a bit awkward to me, especially having a read-only innocuous request cause a table write.

Why don't you just do `select * from community where id = (random() * (select max(id) from community))::int` in a loop? The id of communities is already unique and uniformly distributed, with a fill rate of like 99% (so needing more than one/two queries is exceedingly unlikely) - no new column or modifying query needed.

Reply:
I guess that would make filtered (nswf/local) more difficult, mh.